File: LeetCode/Medium/0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        Map<Integer, List<Character>> map = new HashMap<>();

        map.put(2, List.of('a', 'b', 'c'));
        map.put(3, List.of('d', 'e', 'f'));
        map.put(4, List.of('g', 'h', 'i'));
        map.put(5, List.of('j', 'k', 'l'));
        map.put(6, List.of('m', 'n', 'o'));
        map.put(7, List.of('p', 'q', 'r', 's'));
        map.put(8, List.of('t', 'u', 'v'));
        map.put(9, List.of('w', 'x', 'y', 'z'));

        List<String> result = new ArrayList<>();
        dfs(result, map, digits, 0, new StringBuilder());

        return result;
    }

    private void dfs(List<String> result, Map<Integer, List<Character>> map, String digits, int index, StringBuilder sb) {
        if (index == digits.length()) {
            result.add(sb.toString());
            return;
        }

        for (Character c : map.get(digits.charAt(index) - '0')) {
            dfs(result, map, digits, index + 1, new StringBuilder(sb).append(c));
        }
    }
}

File: LeetCode/Medium/0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.kt
class Solution {
    fun letterCombinations(digits: String): List<String> {
        val map: Map<Int, List<Char>> = mapOf(
            2 to listOf('a', 'b', 'c'),
            3 to listOf('d', 'e', 'f'),
            4 to listOf('g', 'h', 'i'),
            5 to listOf('j', 'k', 'l'),
            6 to listOf('m', 'n', 'o'),
            7 to listOf('p', 'q', 'r', 's'),
            8 to listOf('t', 'u', 'v'),
            9 to listOf('w', 'x', 'y', 'z')
        )

        val result = mutableListOf<String>()

        fun dfs(index: Int, sb: StringBuilder) {
            if (index == digits.length) {
                result.add(sb.toString())
                return
            }

            for (c in map[digits[index] - '0']!!) {
                dfs(index + 1, StringBuilder(sb).append(c))
            }
        }

        dfs(0, StringBuilder())
        return result
    }
}